File: problem-solving/_lc-409-easy-longest-palindrome/1-intuitive.go
/*
https://leetcode.com/problems/longest-palindrome/

Complexity
----------
Time: O(n)
Space: O(n)
*/

func longestPalindrome(s string) int {
    mp := make(map[rune]int)
    for _, c := range s {
        mp[c] += 1
    }

    ans, oddCount := 0, false
    for _, v := range mp {
        if v % 2 == 0 {
            ans += v
        } else {
            ans += v - 1
            oddCount = true
        }
    }

    // We can take only one odd number of times occurring char
    if oddCount {
        ans += 1
    }
    return ans
}

File: problem-solving/_lc-409-easy-longest-palindrome/1-intuitive.py
"""
https://leetcode.com/problems/longest-palindrome/

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def longestPalindrome(self, s: str) -> int:
        count = collections.Counter(s)
        oddCount = False
        ans = 0
        for c in count:
            if count[c] % 2 == 0:
                ans += count[c]
            else:
                ans += count[c] - 1
                oddCount = True
        if oddCount:
            ans += 1
        return ans
